### DISCUSSION TITLE
How to set Numlock

### DISCUSSION THREAD

**Discussion by arjenv** [dosbox-staging/dosbox-staging#4167] [General]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

dosbox-staging, version 0.81.2

### What Operating System are you using?

Linux x86_64

### Whats your question and how can we help?

Is it possible to set (or unset) NUMLOCK in the ON or OFF state at startup? It is not following the state where it is in, I always have to set it to ON.

Fedora 41, dosbox staging 0.81.2

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by johnnovak** [comment]
No, it's not possible to query it at the SDL level. I think those states are entirely managed by the keyboard hardware, not the OS.

This came up before, and maybe there is some low-level OS specific way to get this info from the keyboard, but we didn't wanna go there.

**Comment by johnnovak** [comment]
Oops, closed by fat fingering the close button 😆

**Comment by Torinde** [comment]
Related:
- https://github.com/joncampbell123/dosbox-x/issues/3634#issuecomment-2230394746

> https://github.com/schellingb/dosbox-pure/commit/98d103bb6c9df9dd438e310b14aed16f38afea1f
Pass host keyboard LED state (num-lock, caps-lock, scroll-lock) to emulation if available

https://github.com/joncampbell123/dosbox-x/commit/ac6f282abed641146c7e5473fff09992abf74992

> It turns out the reason Windows XP can sometimes randomly BSOD, expecially in Windows if you Alt+Tab to another application, then Alt+Tab back, is that there is "helpful" code that tries to resynchronize Num+Scroll+Caps state with the guest by modifying the BIOS data area. Those addresses are valid in actual DOS mode but if you do that in the 32-bit environment of the Windows NT kernel you will cause a page fault! Limit that synchronization to DOSBox DOS kernel and real mode.

**Comment by johnnovak** [comment]
> Related:
> 
>     * [Numlock, Caps lock, Scroll lock, Fn lock, Keyboard layout - sync with host joncampbell123/dosbox-x#3634 (comment)](https://github.com/joncampbell123/dosbox-x/issues/3634#issuecomment-2230394746)
> 
> 
> > [schellingb/dosbox-pure@98d103b](https://github.com/schellingb/dosbox-pure/commit/98d103bb6c9df9dd438e310b14aed16f38afea1f)
> > Pass host keyboard LED state (num-lock, caps-lock, scroll-lock) to emulation if available

I have no idea how those work, if they work at all. I don't think is doable via SDL.

**Comment by arjenv** [comment]
Thanks. Let me rephrase the question then: Would it be possible to set the NUMlock state at startup regardless of the current state? 
The thing is that I play some old DOS puzzles in which the numlock is utilised whenever you want to play the next game. It is only a minor nuisance, but I always have to hit the numlock once whenever I have started the game.

**Comment by johnnovak** [comment]
> Thanks. Let me rephrase the question then: Would it be possible to set the NUMlock state at startup regardless of the current state? The thing is that I play some old DOS puzzles in which the numlock is utilised whenever you want to play the next game. It is only a minor nuisance, but I always have to hit the numlock once whenever I have started the game.

Possible, yes. Will we do it? Unlikely. It's not a big deal to hit numlock once at startup 😄 Maybe you can find some old DOS CLI utility that does it (if it works in Staging).

**Comment by arrowgent** [comment]
https://github.com/dosbox-staging/dosbox-staging/issues/1408#issuecomment-981851063

retested again

dosbox-staging 0.82.0
sdl2 2.30.10
linux 6.8.0-51-generic


in dosbox-staging:
start dosbox-staging, no other inputs yet.
numlock **ON** numbers output  789123
press numlock to turn it off
numlock **OFF** functions output (home/end/pgup/pgdn/etc) thus moving the cursor around the prompt.

capslock **OFF** types in lowercase
capslock **ON** types in all uppercase

**Comment by arrowgent** [comment]
im looking if there is an SDL2 enviornment variable eg: `env SDL_do_thing=numlock=on` but i cant seem to find one.

SDL3 might have something

alternatively you could use fedora's `numlockx` program from the repo
usage `numlockx on;dosbox-staging` would work

**Comment by arjenv** [comment]
Thanks for replying. I do not know what SDL2 or SDL3 is. I did manage to install fedora's 'numlockx'. And indeed when I start dosbox as: `numlockx=off;dosbox` it will start with the correct setting. I do not see the light of the numlock go off or on as I type that command.
So it is possible. Thanks for this solution.

For now I just hit the Numlock.  Hope you will program a command setting in the .conf file in the future.

**Comment by LightningStalker** [comment]
sdl2 2.0.20 unfortunately still not working. Hopefully update soon
